server/src/main/java/org/elasticsearch/cluster/coordination/MasterHistory.java

```diff
@@ -0,0 +1,221 @@
+            n
+        );
+    }
+
+    /**
+     * Returns true if for List of master nodes passed in only one non-null node has been master, and the master has switched
+     * from that node to null n times.
+     * @param masters The List of masters to use
+     * @param n The number of times the non-null master must have switched to null
+     * @return True if there has been a single non-null master and it has switched to null n or more times in the given list of masters.
+     */
+    public static boolean hasSameMasterGoneNullNTimes(List<DiscoveryNode> masters, int n) {
+        if (getDistinctMastersSeen(masters).size() != 1) {
+            return false;
+        }
+        boolean seenNonNull = false;
+        int timesMasterHasGoneNull = 0;
+        for (DiscoveryNode master : masters) {
+            if (master != null) {
+                seenNonNull = true;
+            } else if (seenNonNull) {
+                timesMasterHasGoneNull++;
+            }
+        }
+        return timesMasterHasGoneNull >= n;
+    }
+
+    private static Set<DiscoveryNode> getDistinctMastersSeen(List<DiscoveryNode> masters) {
+        return masters.stream().filter(Objects::nonNull).collect(Collectors.toSet());
+    }
+
+    /**
+     * Returns the set of distinct non-null master nodes seen in this history.
+     * @return The set of all non-null master nodes seen. Could be empty
+     */
+    public Set<DiscoveryNode> getDistinctMastersSeen() {
+        List<TimeAndMaster> masterHistoryCopy = getMasterHistoryForLast30Minutes();
+        return masterHistoryCopy.stream().map(timeAndMaster -> timeAndMaster.master).filter(Objects::nonNull).collect(Collectors.toSet());
+    }
+
+    /**
+     * Returns true if a non-null master was seen at any point in the last n seconds, or if the last-seen master was more than n seconds
+     * ago and non-null.
+     * @param n The number of seconds to look back
+     * @return true if the current master is non-null or if a non-null master was seen in the last n seconds
+     */
+    public boolean hasSeenMasterInLastNSeconds(int n) {
+        List<TimeAndMaster> masterHistoryCopy = getMasterHistoryForLast30Minutes();
+        long now = nowSupplier.get();
+        TimeValue nSeconds = new TimeValue(n, TimeUnit.SECONDS);
+        long nSecondsAgo = now - nSeconds.getMillis();
+        return getCurrentMaster() != null
+            || masterHistoryCopy.stream().anyMatch(timeAndMaster -> timeAndMaster.time > nSecondsAgo && timeAndMaster.master != null);
+    }
+
+    /*
+     * This method creates a copy of masterHistory that only has entries from more than 30 minutes before now (but leaves the newest
+     * entry in even if it is more than 30 minutes old).
+     */
+    private List<TimeAndMaster> getMasterHistoryForLast30Minutes() {
+        List<TimeAndMaster> masterHistoryCopy = masterHistory; // In case masterHistory is swapped while we're using it
```

Comment:
Could we get this as parameter and then not have to make this artificial "copy"?

Reply:
Seems the same to me either way, but yeah I can change it to do that.